### PR TITLE
Improve Weapon Enchant Trigger

### PR DESCRIPTION
**Pull request by Vardex**
# Description

With classic release, shaman and rogue use enchant on their weapon, this pull request aim to add stacks information for weapon enchantment (like rogue poison charges) and improve the weapon enchant trigger interface.

image

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Since I don't have the level to use poisons on my rogue I asked some players to test it for me.

# Additional context

It seem like `Crippling Poison` doesn't return stacks with `GetWeaponEnchantInfo()`.

### CODE REVIEW

File: WeakAuras/GenericTrigger.lua [hunk 1]
```diff
@@ -2931,35 +2933,35 @@ do
...
+        WeakAuras.ScanEvents("TENCH_UPDATE");
         WeakAuras.StopProfileSystem("generictrigger");
       end
 
       tenchFrame:SetScript("OnEvent", function(self, event, arg1)
         WeakAuras.StartProfileSystem("generictrigger");
-        if(arg1 == "player") then
-          timer:ScheduleTimer(tenchUpdate, 0.1);
+        if (event == "UNIT_INVENTORY_CHANGED" and arg1 == "player") then
+          tenchUpdate()
```
Comment (mrbuds) [inline WeakAuras/GenericTrigger.lua]: i believe there was a reason to add a 0.1s delay before calling tenchUpdate, why do you remove it ?

File: WeakAuras/Prototypes.lua [hunk 2]
```diff
@@ -4488,47 +4488,53 @@ WeakAuras.event_prototypes = {
...
-      if(trigger.use_enchant and trigger.enchant and trigger.enchant ~= "") then
-        ret = ret .. [[
-          exists = name == ']] .. trigger.enchant .. [[';
-        ]]
-      end
-      return ret;
+        local remaining = expirationTime and expirationTime - GetTime()
+
+        local nameCheck = triggerName == "" or triggerName == name
+        local stackCheck = not triggerStack or stack and stack and stack %s triggerStack
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: this will compare number with nil if trigger.use_stack is false

Comment (mrbuds) [reply WeakAuras/Prototypes.lua]: or stack and stack ??

File: WeakAuras/Prototypes.lua [hunk 3]
```diff
@@ -4488,47 +4488,53 @@ WeakAuras.event_prototypes = {
...
+        local remainingCheck = not triggerRemaining or remaining and remaining %s triggerRemaining
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: same here

File: WeakAuras/Prototypes.lua [hunk 4]
```diff
@@ -4537,11 +4543,29 @@ WeakAuras.event_prototypes = {
...
       },
       {
-        name = "inverse",
-        display = L["Inverse"],
-        type = "toggle",
+        name = "stack",
+        display = L["Stack Count"],
+        type = "number",
+        test = "true",
+        hidden = not WeakAuras.IsClassic()
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: it needs to be disabled too

File: WeakAuras/Prototypes.lua [hunk 5]
```diff
@@ -4537,11 +4543,29 @@ WeakAuras.event_prototypes = {
...
-      }
+      },
+      {
+        name = "showOn",
+        display = L["Show On"],
+        type = "select",
+        values = "weapon_enchant_types",
+        test = "(triggerShowOn == \"showOnActive\" and found) " ..
+        "or (triggerShowOn == \"showOnMissing\" and not found) "  ..
+        "or (triggerShowOn == \"showAlways\")",
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: for readability `triggerShowOn == 'showAlways'`

File: WeakAuras/WeakAuras.lua [hunk 6]
```diff
@@ -3465,6 +3465,27 @@ function WeakAuras.Modernize(data)
     end
   end
 
+  if data.internalVersion < 24 then
+    if data.triggers then
+      for triggerId, triggerData in ipairs(data.triggers) do
+        local trigger = triggerData.trigger
+        if trigger and trigger.type == "status" and trigger.event == "Weapon Enchant" then
+          print("Found a Weapon Enchant")
```
Comment (mrbuds) [inline WeakAuras/WeakAuras.lua]: you left a print here

File: WeakAuras/Prototypes.lua [hunk 7]
```diff
@@ -4501,18 +4501,17 @@ WeakAuras.event_prototypes = {
...
+          expirationTime, duration, name, shortenedName, _, stack, enchantID = WeakAuras.GetMHTenchInfo()
         else
-          expirationTime, duration, name, _, stack = WeakAuras.GetOHTenchInfo()
+          expirationTime, duration, name, shortenedName, _, stack, enchantID = WeakAuras.GetOHTenchInfo()
         end
 
         local remaining = expirationTime and expirationTime - GetTime()
 
-        local nameCheck = triggerName == "" or triggerName == name
+        local nameCheck = triggerName == "" or name and triggerName == name or shortenedName and triggerName == shortenedName or enchantID and triggerName == enchantID
```
Comment (mrbuds) [inline WeakAuras/Prototypes.lua]: tonumber(triggerName) == enchantID

Comment (mrbuds) [reply WeakAuras/Prototypes.lua]: can you add in tooltip that enchantID works too ?

File: WeakAuras/Prototypes.lua [hunk 8]
```diff
@@ -4488,59 +4488,88 @@ WeakAuras.event_prototypes = {
...
         values = "weapon_types",
-        test = "(inverse and not exists) or (not inverse and exists)"
+        test = "true",
+        default = "main",
+        required = true
       },
       {
         name = "enchant",
         display = L["Weapon Enchant"],
+        desc = L["Enchant name or Enchant ID"],
```
Comment (Stanzilla) [inline WeakAuras/Prototypes.lua]:
```suggestion
        desc = L["Enchant Name or ID"],
```

ID(s) if it supports more than one.

Comment (mrbuds) [reply WeakAuras/Prototypes.lua]: Weapon Enchantment Name or ID

**Pull request merged**